### ISSUE TITLE
Gopls Binary Location

### ISSUE THREAD

**Issue by jvonfricken** [GwynethLlewelyn/Go.novaextension#2]
I've been able to get gopls _barely_ working with nova. The issue is that the gopls binary needs to be fetched statically from the go/bin directory. Example: `/go/bin/gopls`.  I'd be happy to make a PR with my changes, if you'd like. 

<img width="1501" alt="Screen Shot 2020-09-17 at 11 40 18 AM" src="https://user-images.githubusercontent.com/26029263/93494218-c6cb6980-f8da-11ea-884e-f9979635b076.png">

Although, it's still extremely unstable, and I'm not sure if it's Nova's fault. But, I'll keep working away at it.

**Comment by tooolbox** [comment]
First off, I appreciate your work :)

Now, without actually investigating, I question the phrase "fetched statically".  The [prettier extension](https://github.com/alexanderweiss/nova-prettier) seems to be using the system-installed prettier just fine, so if `gopls` is in your path why wouldn't that just work?  Is it a matter of what arguments are passed to it, perhaps?  Maybe the [VSCode extension](https://github.com/golang/vscode-go) could shed some light on this?

**Comment by jvonfricken** [comment]
Yeah definitely.

That was not the clearest I could have phrased that lol. In the README.md, the author says to copy the gopls binary into the application path of Nova. However, I found success by changing the gopls path in `main.js` to be the `go/bin` path.

Is that more clear?

**Comment by jvonfricken** [comment]
Additionally, I forgot to add that even if you do this, once you actually select anything from the autocomplete menu, Nova immediately crashes, with an inscrutable crash.log -- at least to me --

**Comment by GwynethLlewelyn** [comment]
Both suggestions sound good!

I _did_ try to use a system-wide (well, `$PATH`-based) `gopls`, but I had a few issues with it...

1. The first iterations of the Nova beta simply ignored everything _outside_ the extension folder. I have no idea if that just was me being stupid, an incomplete/erroneous implementation by Panic, or an actual security issue (i.e. deliberately preventing people to launch whatever they wanted _outside_ the extension folder...).
2.  _Prettier_, as well as a few others (_CSLint_ comes to mind), seem to work fine with system-wide versions of those applications, so I guess it would work with `gopls`, too. Ironically, on the last few Nova releases, all those extensions stopped working on _my_ Mac – but I admit it might be a problem with my own setup. However, while people working on websites, frontends, etc. are likely to have those utilities installed (or at least _know_ how to install them!), `gopls` is another story – even if you're a seasoned Go programmer, it's highly likely you will _not_ have it installed – and the installation is not trivial, although I guess that at least a Go developer will have a working Go environment...

That said, since `gopls` is too big to be bundled with the extension anyway, I'm going to try using the 'system-wide' version, or, failing that, the one (possibly) under `./go/bin`.

On the other hand, I'm astonished that you managed to get autocomplete to _work_! That's a huge surprise to me, since I didn't see anything happening... ok, sure, so, Nova crashes, but it's still a step in the right direction, I guess...

**Comment by jvonfricken** [comment]
It wouldn't surprise me that there was some inconsistency with path access in the beta versions of Nova. Apple's APIs are making it harder and harder to access the filesystem on macOS. 

I think it would make the most sense to follow VSCode with regards to gopls. The application itself will pop a notification if gopls is missing, and then offer to pull it down for you via the notification. I don't know if the Nova extension library will allow that, though. 

I have my fingers crossed that in an upcoming update, the Panic folks will have a patch that will fix the crashing, but until then I'm pretty much blocked on any further development. Let me know if you have anymore luck than I did!

**Comment by jfieber** [comment]
@jvonfricken — With what version of `gopls`, macOS and Nova did you find this partial success?

**Comment by jvonfricken** [comment]
@jfieber I'm using `gopls` v0.5.0, and version 10.15.6  macOS.

**Comment by jfieber** [comment]
@GwynethLlewelyn After some poking around, I've discovered that `"Go"` in `syntaxes` in the client options needs to be lower case [here](https://github.com/GwynethLlewelyn/Go.novaextension/blob/de3ec439677cd8d185ae5faff35214fa3b337add/Scripts/main.js#L21). Without that change, I was seeing `gopls` start up, but Nova doesn't actually use it. Now hover documentation works for me, though shows as bare Markdown:

<img width="1054" alt="image" src="https://user-images.githubusercontent.com/113956/94375760-c6955000-00ca-11eb-883c-ccbe7b52ccdd.png">

I do get some auto-complete popups, but the suggestions are…a bit strange, offering mostly invalid suggestions, and just a handful of valid ones.  When there is a sensible suggestion, such as for `fmt.er`:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/113956/94376450-62c15600-00cf-11eb-9d09-2d26b6088e9b.png">

Selecting the completion yields the not-so-helpful result:

<img width="287" alt="image" src="https://user-images.githubusercontent.com/113956/94376475-8d131380-00cf-11eb-9450-319536a2068f.png">

With respect to the where `gopls` is located, I'm not seeing any different behavior based on where it is loaded from.

This Nova build 204146, macOS 10.15.6 and `gopls` v0.5.0.

**Comment by jvonfricken** [comment]
@jfieber Thanks for bringing up the syntax needing to be `go`. That was the one other thing I had done, but forgotten. 

As far as I can tell, Nova is not respecting the variable substitutions coming back as a response from `gopls`.

**Comment by tooolbox** [comment]
> As far as I can tell, Nova is not respecting the variable substitutions coming back as a response from gopls.

Possibly the extension needs to transform the substitutions into what Nova expects?

**Comment by jvonfricken** [comment]
Agreed. The Typescript extension does that, I believe. I hope the Panic folks publish some good detailed docs about how their LSP functionality works. It would make the translation shim much easier to implement.

**Comment by jfieber** [comment]
I believe #6 should address the issue of where to find `gopls`.

**Comment by jvonfricken** [comment]
@jfieber That's great!

Seems like the Nova 2 update fixed some of the issues with `gopls`. However, it's far from stable. But, it's promising that Panic is making progress at such a rapid clip.

**Comment by jfieber** [comment]
@jvonfricken I do have updates I hope to get cleaned up in the coming weekend.  With Nova 2 we can now pass in custom parameters in the client/server initialization, so we can turn [these knobs](https://github.com/golang/tools/blob/master/gopls/doc/settings.md). It currently requires completely restarting the LanguageClient and gopls when they change, but that should be addressed in an upcoming Nova update.  Promises and async/await weren't a thing last time I worked in JavaScript, so I updated some of my JS knowledge and gopls restarts are pretty reliable now.

Essential for my use, I have jump to definition/type definition/implementation(s)/references along with a jump stack that you can unwind to get back to where you jumped from. The built in jump-to-definition now uses gopls correctly in Nova 2 and I expect the others will come as Nova built-in, since they are all part of the core protocol. As such, I'm not likely to spend too much effort refining my implementations.

There continues to be situations where gopls and Nova appear to get out of sync with each other with respect to the state of the workplace, and I haven't pinned down a particular situation to reliably reproduce it. Completions sometimes fail. Same with signature help. Format and Organize Imports (my implementations) intermittently don't work right, reporting no changes to make. They may all be related. In Nova 2, the built in code action support offers up an Organize Imports option, but it doesn't appear to work at all.

**Comment by GwynethLlewelyn** [comment]
Sorry for my absence!... Belated Happy New Year to you all!

Ok, I need to think this over, I guess there are a lot of new things requiring fixing.

One thing I'll add is that these days you can easily get `gopls` with `brew install gopls`. So there! End of problems. Those who hate Homebrew for some reason can always compile it on their own.

**Comment by jfieber** [comment]
Alas, gopls reliably [crashes Nova](https://github.com/jfieber/GoBang.novaextension). I send in a new crash dump on every Nova update.  There is a workaround by disabling the _Code actions indicator_ setting in Nova. I also find that Nova and gopls often get out-of-sync with each other. On the surface, it looks like Nova is doing doing the Right Thing, but I still need to compare RPC logs of another gopls using editor.

**Comment by GwynethLlewelyn** [comment]
... I should have read your comment _before_ the last extension submit 🤣 That's fine, we can always change the text on the README.

I've been twisting my brain trying to figure out _why_ Nova crashes; my guess is that some relevant structures have reserved memory but not properly released it; I _think_ that this is mostly because the Microsoft code used by @apexskier on his TypeScript extension seems to initialise things in a slightly different way...

**Comment by GwynethLlewelyn** [comment]
... BTW, for @jvonfricken — under Nova 4.3, `gopls` can simply be installed using `brew install gopls` and Nova is dealing much better with figuring out paths...

Or, of course, you can do a `go get` and just run it from there.

What amazes me is that you manage to get proper formatting on the tooltips... but that's a different issue!

**Comment by GwynethLlewelyn** [comment]
I _think_ I've fixed the issue for @jvonfricken — i.e. the path issue for `gopls`. The other sub-issues mentioned in this thread have been partially addressed 😁 so they're technically open. I'm currently working on 'activating' some of the cool features from @jfieber's merged code. It's been an education so far!
